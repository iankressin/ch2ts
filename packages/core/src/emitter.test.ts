import { describe, expect, it } from "vitest";
import {
  emit,
  map,
  parse,
  type EmissionOptions,
  type MappingOptions,
} from "./index.js";

describe("emitter (ts-morph)", () => {
  const baseMap: MappingOptions = {
    int64As: "bigint",
    decimal: "string",
    datetimeAs: "string",
    camelCase: true,
  };

  it("emits interface with JSDoc and branded types", () => {
    const ddl = `
      /* block comment */
      CREATE TABLE db.events (
        ip IPv4 COMMENT 'client ip',
        id UInt64,
        name String -- trailing
      );
    `;
    const ast = parse(ddl);
    const mapped = map(ast, baseMap);
    const out = emit(mapped, { emitZod: false } satisfies EmissionOptions);
    expect(out).toMatchSnapshot();
  });

  it("emits zod schemas for simple types", () => {
    const ddl = `CREATE TABLE t (a UInt64, b Nullable(String), c Array(UInt32));`;
    const ast = parse(ddl);
    const mapped = map(ast, baseMap);
    const out = emit(mapped, { emitZod: true } satisfies EmissionOptions);
    expect(out).toMatchSnapshot();
  });
});
