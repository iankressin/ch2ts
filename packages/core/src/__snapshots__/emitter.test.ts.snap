// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emitter (ts-morph) > emits interface with JSDoc and branded types 1`] = `
"/*
 * Generated by @ch2ts/core
 * Options: {"emitZod":false}
 */
export interface Events {
  /** Original: IPv4 â€” client ip */
  ip: string;
  /** Original: UInt64 */
  id: bigint;
  /** Original: String */
  name: string;
}
"
`;

exports[`emitter (ts-morph) > emits zod schemas for simple types 1`] = `
"/*
 * Generated by @ch2ts/core
 * Options: {"emitZod":true}
 */
import { z } from "zod";

export interface T {
  /** Original: UInt64 */
  a: bigint;
  /** Original: Nullable(String) */
  b: string | null;
  /** Original: Array(UInt32) */
  c: number[];
}

export const TSchema = z.object({ a: z.bigint(), b: z.string().nullable(), c: z.array(z.number()) });
"
`;
